keps/sig-apps/4212-declarative-node-maintenance/README.md

```diff
@@ -0,0 +1,2300 @@
+  evacuation from the first node.
+- `.status.drainStatus.podsEvacuating`, to indicate how many pods are being evacuated from the first node.
+  These are the pods that have matching Evacuation objects. 
+- To keep track of the entire maintenance the controller will reconcile a `Drained` condition and
+  set it to true if all pods pending evacuation/termination have terminated on all target nodes
+  when drain is requested by the maintenance object.
+- NodeMaintenance condition or annotation can be set on the node object to advertise the current
+  phase of the maintenance.
+
+#### Supported Stage Transitions
+
+The following transitions should be validated by the API server.
+
+- Idle -> _Deletion_
+  - Planning a maintenance in the future and canceling/deleting it without any consequence.
+- (Idle) -> Cordon -> (Complete) -> _Deletion_.
+  - Make a set of nodes unschedulable and then schedulable again.
+  - The complete stage will always be run even without specifying it.
+- (Idle) -> (Cordon) -> Drain -> (Complete) -> _Deletion_.
+  - Make a set of nodes unschedulable, drain them, and then make them schedulable again.
+  - Cordon and Complete stages will always be run, even without specifying them.
+- (Idle) -> Complete -> _Deletion_.
+  -  Make a set of nodes schedulable.
+
+The stage transitions are invoked either manually by the cluster admin or by a higher-level
+controller. For a simple drain, cluster admin can simply create the NodeMaintenance with
+`stage: Drain` directly.
+
+### DaemonSet Controller
+
+The DaemonSet workloads should be tied to the node lifecycle because they typically run critical
+workloads where availability is paramount. Therefore, the DaemonSet controller should respond to
+the Evacuation only if there is a NodeMaintenance happening on that node and the DaemonSet is in
+the `drainTargets`. For example, if we observe the following NodeMaintenance:
+
+```yaml
+apiVersion: v1alpha1
+kind: NodeMaintenance
+...
+status:
+  nodeStatuses:
+    - nodeRef:
+        name: six
+      drainTargets:
+        - podPriority: 2147483647
+          podType: Default
+        - podPriority: 5000
+          podType: DaemonSet
+  ...
+```
+
+To fulfil the Evacuation API, the DaemonSet controller should register itself as a controller
+evacuator. To do this, it should ensure that the following annotation is present on its own pods.
+
+```yaml
+evacuation.coordination.k8s.io/priority_daemonset.apps.k8s.io: "10000/controller"
+```
+
+The controller should respond to the Evacuation object when it observes its own class
+(`daemonset.apps.k8s.io`) in `.status.activeEvacuatorClass`.
+
+For the above node maintenance, the controller should not react to Evacuations of DaemonSet pods
+with a priority greater than 5000. This state should not normally occur, as Evacuation requests
+should be coordinated with NodeMaintenance. If it does occur, we should not encourage this flow by
```

Comment:
It can be used if you implement a custom Evacuator. I think it does not make sense for the DaemonSet controller to implement it separately without the NodeMaintenance ATM.

Reply:
It would have been nice if we would be able to evict DaemonSet Pods independent of the NodeMaintenance API if the NodeMaintenance API cannot be used.

Example, I could imagine that NodeMaintenance API will not be compatible with what we want to be able to do in Cluster API, so it would have been just nice to use it without the NodeMaintenance API.

(Context: https://github.com/kubernetes-sigs/cluster-api/blob/main/docs/proposals/20240930-machine-drain-rules.md)